### PR TITLE
Add start/stop api

### PR DESCRIPTION
**Pull request by osdnk**
## Motivation
Handle start Api like in RN core

## Changes
 - Add `always` node - `final` node (like `props`) which evaluate its children
 - Add start and stop function which attach and detach node as a child of value we wish to "override" with animation
 - Change way of evaluating node in topological order by final nodes
 - Add setValue and evaluateOnce

### CODE REVIEW

File: android/src/main/java/com/swmansion/reanimated/nodes/Node.java [hunk 1]
```diff
@@ -93,7 +95,7 @@ protected final void forceUpdateMemoizedValue(T value) {
     markUpdated();
   }
 
-  private static void findAndUpdateNodes(Node node, Set<Node> visitedNodes) {
+  private static void findAndUpdateNodes(Node node, Set<Node> visitedNodes, Deque<FinalNode> finalsToBeUpdated) {
```
Comment (kmagiera) [inline android/src/main/java/com/swmansion/reanimated/nodes/Node.java]: rename -> finalNodesQueue

Also I think you want to just use Stack instead of Deque

File: ios/Nodes/REANode.m [hunk 2]
```diff
@@ -107,26 +107,33 @@ - (void)markUpdated
...
+    [finalsToBeUpdated addObject:node];
   }
 }
 
 + (void)runPropUpdates:(REAUpdateContext *)context
 {
   NSMutableSet<REANode *> *visitedNodes = [NSMutableSet new];
   for (NSUInteger i = 0; i < context.updatedNodes.count; i++) {
-    [self findAndUpdateNodes:context.updatedNodes[i] withVisitedSet:visitedNodes];
+    NSMutableArray<REANode *> * finalsToBeUpdated = [NSMutableArray new];
```
Comment (kmagiera) [inline ios/Nodes/REANode.m]: shouldn't we have just one stack outside of the above `for` loop?

File: src/animations/TimingAnimation.js [hunk 3]
```diff
@@ -48,4 +48,12 @@ export default class TimingAnimation extends Animation {
   stop() {
     // this._finished && this._finished.setValue(1);
   }
+  static persistenceStateValue = 'frameTime';
+  static getDefaultState() {
+    return {
+      finished: new AnimatedValue(0),
+      time: new AnimatedValue(0),
+      frameTime: new AnimatedValue(0),
```
Comment (kmagiera) [inline src/animations/TimingAnimation.js]: This seems to be the same for all animation types, no?

File: src/animations/backwardsCompatibleWrapper.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    return {
+      __state: {
+        // for sequencing
+        finished: _state.finished,
+        node: createSetNode,
+        disableDetaching: () => (enabledDetaching = false),
+        attachToVal: an => an.__addChild(_value),
+        detachFromVal: an => an.__removeChild(_value),
+      },
+      start: _returnMethod => {
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleWrapper.js]: why prefix with `_` – it is normally used for class member methods

File: src/animations/backwardsCompatibleWrapper.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        if (isStarted) {
+          console.warn(
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleWrapper.js]: Is that the same behavior as for in the original animated implementation?

File: src/animations/backwardsCompatibleWrapper.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+} from '../base';
+import { default as Clock } from '../core/AnimatedClock';
+import { default as Value } from '../core/AnimatedValue';
+
+export default function backwardsCompatibleWrapper(node, AnimationClass) {
+  return (clock, state, config) => {
+    if (config !== undefined) {
+      return node(clock, state, config);
+    }
+    const _value = clock;
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleWrapper.js]: please don't use `_` prefix here either

File: src/animations/backwardsCompatibleWrapper.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  cond,
+  set,
+  startClock,
+  stopClock,
+} from '../base';
+import { default as Clock } from '../core/AnimatedClock';
+import { default as Value } from '../core/AnimatedValue';
+
+export default function backwardsCompatibleWrapper(node, AnimationClass) {
+  return (clock, state, config) => {
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleWrapper.js]: as this arrow method seems pretty long it would be better to extract it to the top level and use `function` to call it somehow, then use if the below condition fails

This will also give us an opportunity to free up variable names for clock etc

File: src/animations/backwardsCompatibleWrapper.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+      ]);
+    const wrappedNode = node(newClock, _state, _config);
+    const shouldStopClockInNexFrame = new Value(0); // needed for seq
+    const createSetNode = () =>
+      set(
+        _value,
+        block([
+          ps && [
+            initPersCache(),
+            call([_state[ps]], p => (cachedValue = p[0])),
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleWrapper.js]: it doesn't seem like you actually use `cachedValue` anywhere except from in `initPersCache` which gets called before you update it

File: src/animations/backwardsCompatibleWrapper.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+          ],
+          cond(shouldStopClockInNexFrame, [
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleWrapper.js]: next

File: src/animations/backwardsCompatibleWrapper.js [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+          wrappedNode,
+          cond(_state.finished, [
+            call([], () => {
+              finished = 1;
+              if (enabledDetaching) {
+                alwaysNode.__removeChild(_value);
+              }
+              returnMethod && returnMethod({ finished: true });
+            }),
+            set(shouldStopClockInNexFrame, 1),
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleWrapper.js]: why can't wee simply stop the clock here instead of scheduling it for the next frame?

File: src/animations/TimingAnimation.js (hunk 3 above)
Comment (osdnk) [reply src/animations/TimingAnimation.js]: No, it's not

File: src/animations/backwardsCompatibleWrapper.js (hunk 10 above)
Comment (osdnk) [reply src/animations/backwardsCompatibleWrapper.js]: It's connected with sequencing. 
Take a look: if we stop clock now, `cond` in sequnce.js will not be evaluated after setting state to `finished` and it will won't trigger next animation.

File: src/Animated.js [hunk 11]
```diff
@@ -48,12 +33,13 @@ const Animated = {
   ...derived,
 
   // animations
-  decay: backwardsCompatibleAnim(decay, DecayAnimation),
-  timing: backwardsCompatibleAnim(timing, TimingAnimation),
-  spring: backwardsCompatibleAnim(spring, SpringAnimation),
+  decay: backwardsCompatibleWrapper(decay, DecayAnimation),
```
Comment (kmagiera) [inline src/Animated.js]: can we rename this so that it contains "Anim" or sth that would indicate it relates to animation

File: src/utils.js [hunk 12]
```diff
@@ -16,3 +16,24 @@ export function adapt(v) {
 export function val(v) {
   return v && v.__getValue ? v.__getValue() : v || 0;
 }
+
+export function evaluateOnce(node, children = []) {
```
Comment (kmagiera) [inline src/utils.js]: should go under `derived/` somewhere

File: src/animations/backwardsCompatibleWrapper.js [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+  always,
+  block,
+  call,
+  clockRunning,
+  cond,
+  set,
+  startClock,
+  stopClock,
+} from '../base';
+import { delay } from '../derived';
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleWrapper.js]: let's move delay to separate PR

File: src/animations/timing.js [hunk 14]
```diff
@@ -7,22 +7,24 @@ import {
...
   greaterOrEq,
+  defined,
 } from '../base';
 
 export default function timing(clock, state, config) {
   const lastTime = cond(state.time, state.time, clock);
   const frameTime = add(state.frameTime, sub(clock, lastTime));
 
-  const progress = config.easing(divide(state.frameTime, config.duration));
+  const duration = add(config.duration, config.delay);
```
Comment (kmagiera) [inline src/animations/timing.js]: revert

File: src/core/AnimatedNode.js [hunk 15]
```diff
@@ -2,6 +2,8 @@ import ReanimatedModule from '../ReanimatedModule';
 
 const UPDATED_NODES = [];
 
+export const initializedNodes = new Set();
```
Comment (kmagiera) [inline src/core/AnimatedNode.js]: I suggest that we keep this object internal to the current module and only export a method returning its size.

We can call it:
```
getNumberOfInitializedNodes_onlyForTests()
```

File: src/utils.js (hunk 12 above)
Comment (osdnk) [reply src/utils.js]: basically we dont need it now

File: src/utils.js [hunk 16]
```diff
@@ -1,7 +1,6 @@
 import AnimatedBlock from './core/AnimatedBlock';
 import AnimatedNode from './core/AnimatedNode';
 import AnimatedValue from './core/AnimatedValue';
-
```
Comment (kmagiera) [inline src/utils.js]: lets keep that empty line here

File: src/derived/evaluateOnce.js [hunk 17]
```diff
@@ -0,0 +1,28 @@
+import AnimatedValue from '../core/AnimatedValue';
+import { block, call, always, cond, set } from '../base';
+export function evaluateOnce(node, children = []) {
```
Comment (kmagiera) [inline src/derived/evaluateOnce.js]: add empty line above

File: src/derived/evaluateOnce.js [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+  let _children;
```
Comment (kmagiera) [inline src/derived/evaluateOnce.js]: let's not use underscore as prefix for inner variables. It should only be used for object member variables to indicate that the variable should be considered "private"

File: README.md [hunk 19]
```diff
@@ -463,7 +470,7 @@ Creates a color node in RGBA format. Where first three input nodes should have i
 
 The returned node can be mapped to view properties that represents color (e.g. [`backgroundColor`](https://facebook.github.io/react-native/docs/view-style-props.html#backgroundcolor)).
 
-<!-- Anims -->
+## Animations
```
Comment (kmagiera) [inline README.md]: remove

File: android/src/main/java/com/swmansion/reanimated/nodes/Node.java [hunk 20]
```diff
@@ -93,7 +94,7 @@ protected final void forceUpdateMemoizedValue(T value) {
     markUpdated();
   }
 
-  private static void findAndUpdateNodes(Node node, Set<Node> visitedNodes) {
+  private static void findAndUpdateNodes(Node node, Set<Node> visitedNodes, Stack<FinalNode> finalNodesQueue) {
```
Comment (kmagiera) [inline android/src/main/java/com/swmansion/reanimated/nodes/Node.java]: It's a stack, not queue. Maybe just rename to "finalNodes"

File: ios/Nodes/REAAlwaysNode.m [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+
+@implementation REAAlwaysNode
+{
+  NSNumber * _nodesToBeEvaluated;
+}
+
+- (instancetype)initWithID:(REANodeID)nodeID config:(NSDictionary<NSString *,id> *)config
+{
+    if ((self = [super initWithID:nodeID config:config])) {
+        _nodesToBeEvaluated = config[@"what"];
```
Comment (kmagiera) [inline ios/Nodes/REAAlwaysNode.m]: it's just one node, not many. Rename to "_nodesToBeEvaluated"

File: ios/Nodes/REAAlwaysNode.m [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+    }
+    return self;
+}
+
+- (id)value
```
Comment (kmagiera) [inline ios/Nodes/REAAlwaysNode.m]: I think you should override `evaluate` here instead of `value`

`value` is a method that should not be overwritten (although there is no way in objc to prevent that). It calls into `evaluate` and memoizes the result so that it is guaranteed we "evaluate" at most once per frame

File: ios/Nodes/REANode.m [hunk 23]
```diff
@@ -107,26 +107,33 @@ - (void)markUpdated
...
-      [self findAndUpdateNodes:child withVisitedSet:visitedNodes];
-    }
+    [finalNodesQueue addObject:node];
   }
 }
 
 + (void)runPropUpdates:(REAUpdateContext *)context
 {
   NSMutableSet<REANode *> *visitedNodes = [NSMutableSet new];
+  NSMutableArray<REANode *> * finalNodesQueue = [NSMutableArray new];
```
Comment (kmagiera) [inline ios/Nodes/REANode.m]: let's not use "queue" to name a stack

File: src/animations/backwardsCompatibleAnimWrapper.js [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+      if (isStarted) {
+        returnMethod && returnMethod({ finished: false });
+        return;
+      }
+      isStarted = true;
+      alwaysNode = always(
+        set(
+          value,
+          block([
+            cond(currentState.finished, set(currentState.finished, 0)),
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleAnimWrapper.js]: are you sure about this one? What's the purpose of `finished` if we just always set it to 0.

I think we should only expect `finished` to be updated by the animation node

File: src/animations/backwardsCompatibleAnimWrapper.js [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+            cond(clockRunning(newClock), 0, [
+              set(newValue, value),
+              startClock(newClock),
+            ]),
+            node(newClock, currentState, config),
+            cond(currentState.finished, [
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleAnimWrapper.js]: but you set it to `0` few lines earlier

File: src/animations/backwardsCompatibleAnimWrapper.js [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+  clockRunning,
+  cond,
+  set,
+  startClock,
+  stopClock,
+} from '../base';
+import { default as Clock } from '../core/AnimatedClock';
+import { default as Value } from '../core/AnimatedValue';
+
+function backwardsCompatibleInvoke(node, AnimationClass, value, config) {
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleAnimWrapper.js]: Is there still a point in having that `AnimationClass` defined for each type? It looks like this code can be unified and just be moved here instead of having it duplicated for each animation type

File: src/animations/backwardsCompatibleAnimWrapper.js [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+          ])
+        )
+      );
+      returnMethod = currentReturnMethod;
+      alwaysNode.__addChild(value);
+    },
+    stop: () => {
+      isStarted = false;
+      alwaysNode.__removeChild(value);
+      returnMethod && returnMethod({ finished: false });
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleAnimWrapper.js]: I'm worried that under certain circumstances this could be called twice (here and in the `call` body defined above). I'd prefer if for `stop` method we just do `set(finished, 1)` then we could only have a single place that would handle "detaching"

File: src/Animated.test.js [hunk 28]
```diff
@@ -0,0 +1,102 @@
+import { getNumberOfInitializedNodes_forTestsOnly as numberOfNodes } from './core/AnimatedNode';
```
Comment (kmagiera) [inline src/Animated.test.js]: Looks like this isn't really necessary. I think what we could do instead is to have our `ReanimatedModule` mock to verify we have equal number of calls to `createNode` and `dropNode`

File: src/Animated.test.js [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+      after === before &&
+      during > after &&
+      initial === 0 &&
+      before === 2;
+    if (pass) {
+      return {
+        message: () =>
+          `expected ${
+            animation.name
+          }'s nodes to be attached and detached correctly`,
```
Comment (kmagiera) [inline src/Animated.test.js]: Do we need "message" if `"pass: true"`? Is it being shown at all? I though the message is only needed when test fail so that we can tell what was wrong

File: src/Animated.test.js [hunk 30]
```diff
@@ -0,0 +1,102 @@
...
+        message: () =>
+          `expected ${
+            animation.name
+          }'s nodes not to be attached and detached correctly`,
+        pass: false,
+      };
+    }
+  },
+});
+it('fails if timing does attach nodes correctly', () => {
```
Comment (kmagiera) [inline src/Animated.test.js]: It sounds a bit unintuitive. Why we expect it to fail if it does stuff correctly?

File: src/Animated.test.js [hunk 31]
```diff
@@ -0,0 +1,102 @@
+import { getNumberOfInitializedNodes_forTestsOnly as numberOfNodes } from './core/AnimatedNode';
+import Animated, { Easing } from './Animated';
+import AnimatedAlways from './core/AnimatedAlways';
+
+jest.mock('./ReanimatedEventEmitter');
+jest.mock('./ReanimatedModule');
+
+const { Value, timing, spring, decay } = Animated;
+expect.extend({
+  toAttachNodesProperly(animation) {
```
Comment (kmagiera) [inline src/Animated.test.js]: This does not seem like a legitimate usecase of `expect.extend` this could be just a method that takes animation object and then evaluates some assertions.

File: src/animations/backwardsCompatibleAnimWrapper.js (hunk 26 above)
Comment (osdnk) [reply src/animations/backwardsCompatibleAnimWrapper.js]: Each animation has different state

File: src/animations/backwardsCompatibleAnimWrapper.js (hunk 24 above)
Comment (osdnk) [reply src/animations/backwardsCompatibleAnimWrapper.js]: I made it to be the same like in RN, but as it's not crucial, gonna drop it

File: README.md (hunk 19 above)
Comment (osdnk) [reply README.md]: Can't get it

File: src/animations/backwardsCompatibleAnimWrapper.js (hunk 27 above)
Comment (osdnk) [reply src/animations/backwardsCompatibleAnimWrapper.js]: done

File: ios/Nodes/REANode.m [hunk 32]
```diff
@@ -107,26 +107,34 @@ - (void)markUpdated
...
+  NSMutableArray<REANode *> *finalNodes = [NSMutableArray new];
   for (NSUInteger i = 0; i < context.updatedNodes.count; i++) {
-    [self findAndUpdateNodes:context.updatedNodes[i] withVisitedSet:visitedNodes];
+    [self findAndUpdateNodes:context.updatedNodes[i]
+              withVisitedSet:visitedNodes
+              withFinalNodes:finalNodes];
+  }
+  while (finalNodes.count > 0) {
+    // NSMutableArray used for stack implementation
+    [(id)[finalNodes lastObject] update];
```
Comment (kmagiera) [inline ios/Nodes/REANode.m]: `(id)` cast shouldn't be needed

File: src/Animated.test.js [hunk 33]
```diff
@@ -0,0 +1,106 @@
...
+const { Value, timing, spring, decay } = Animated;
+describe('Reanimated backward compatible API', () => {
+  beforeEach(() => {
+    let nodesCreated = 0;
+    ReanimatedModule.createNode = () => nodesCreated++;
+    ReanimatedModule.dropNode = () => nodesCreated--;
+    ReanimatedModule.getNumberOfNodes = () => nodesCreated;
+  });
+
+  const checkIfAttachAndDetachNodesProperly = jest.fn(animation => {
```
Comment (kmagiera) [inline src/Animated.test.js]: I don't think there is a point in using `jest.fn` for this method. `js.fn` should only be used for "stubbed" or "mocked" methods that you need later to assert on if they got called and with what arguments etc. This is not the case here

File: src/Animated.test.js [hunk 34]
```diff
@@ -0,0 +1,106 @@
...
+jest.mock('./ReanimatedEventEmitter');
+jest.mock('./ReanimatedModule');
+jest.mock('./derived/evaluateOnce');
+
+const { Value, timing, spring, decay } = Animated;
+describe('Reanimated backward compatible API', () => {
+  beforeEach(() => {
+    let nodesCreated = 0;
+    ReanimatedModule.createNode = () => nodesCreated++;
+    ReanimatedModule.dropNode = () => nodesCreated--;
```
Comment (kmagiera) [inline src/Animated.test.js]: this does not reflect the number of created nodes if you decrement it. Maybe rename to `activeNodes` or just `numberOfNodes`

File: src/core/AnimatedValue.js [hunk 35]
```diff
@@ -14,21 +16,44 @@ export default class AnimatedValue extends AnimatedNode {
...
+
   __onEvaluate() {
     if (this.__inputNodes && this.__inputNodes.length) {
       this.__inputNodes.forEach(val);
     }
     return this._value + this._offset;
   }
 
+  __removeChild(child) {
+    super.__removeChild(child);
```
Comment (kmagiera) [inline src/core/AnimatedValue.js]: no need to override this method

File: src/core/AnimatedValue.js [hunk 36]
```diff
@@ -14,21 +16,44 @@ export default class AnimatedValue extends AnimatedNode {
...
+  }
+
   _updateValue(value) {
     this._value = value;
     this.__forceUpdateCache(value);
   }
 
+  setValue(value) {
+    evaluateOnce(set(this, value), this);
```
Comment (kmagiera) [inline src/core/AnimatedValue.js]: we may want to "detach animation" here too as that's how its done in `Animated`

File: src/derived/evaluateOnce.js [hunk 37]
```diff
@@ -0,0 +1,33 @@
...
+ */
+export function evaluateOnce(node, children = [], callback) {
+  if (!Array.isArray(children)) {
+    children = [children];
+  }
+  const done = new AnimatedValue(0);
+  const evalNode = cond(
+    done,
+    0,
+    block([
```
Comment (kmagiera) [inline src/derived/evaluateOnce.js]: don't think you should need to use `block` here. Just pass in the array and it should get converted to block automatically.

Would be even better if instead of using `block` we just have that `call` node directly here and have `node` and `set(done, 1)` as input arguments for the `call`

File: src/Animated.test.js [hunk 38]
```diff
@@ -0,0 +1,106 @@
...
+    const transX = new Value(0);
+
+    const initial = ReanimatedModule.getNumberOfNodes();
+    const v = new AnimatedNode({ type: 'sampleView', value: 0 });
+    transX.__addChild(v);
+    const before = ReanimatedModule.getNumberOfNodes();
+    const anim = animation.node(transX, animation.config);
+    anim.start();
+    const during = ReanimatedModule.getNumberOfNodes();
+    anim.__value_testOnly.__setAnimation(null, true);
```
Comment (kmagiera) [inline src/Animated.test.js]: IMO this test does not verify much. You could just call enough methods here to make the test pass. How can we be sure these methods are going to be called in real life scenario (especially `__value_testOnly.setAnimation(null, true)`)

We need a test that verifies:
 1) that nodes get detached correctly if we have mounted component that uses value that we start an animation for using imperative API (we don't need to test all possible animation types for that)
 2) nodes get detached correctly if that component get unmounted while the animation is running and that finished callback will get called
 3) nodes get detached correctly if we have two components with a single value we start an animation for, then only one these components get unmounted (the value should stay attached)

File: src/animations/backwardsCompatibleAnimWrapper.js [hunk 39]
```diff
@@ -0,0 +1,96 @@
...
+  let isDone = false;
+  let wasStopped = false;
+  return {
+    start: currentReturnMethod => {
+      if (isStarted) {
+        returnMethod && returnMethod({ finished: false });
+        return;
+      }
+      if (isDone) {
+        console.warn('Animation has been finished before');
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleAnimWrapper.js]: how is this different comparing to what we check above? If `isStarted` is true and we start again shouldn't we warn too? Instead we call the finished callback – is that consistent?

File: src/animations/backwardsCompatibleAnimWrapper.js [hunk 40]
```diff
@@ -0,0 +1,96 @@
...
+import { default as Clock } from '../core/AnimatedClock';
+import { default as Value } from '../core/AnimatedValue';
+import { evaluateOnce } from '../derived/evaluateOnce';
+
+function backwardsCompatibleInvoke(node, AnimationClass, value, config) {
+  let returnMethod;
+  const newValue = new Value(0);
+  const newClock = new Clock();
+  const currentState = AnimationClass.getDefaultState();
+  currentState.position = newValue;
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleAnimWrapper.js]: why `position` is not a part of "defaut state" then?

File: src/animations/backwardsCompatibleAnimWrapper.js [hunk 41]
```diff
@@ -0,0 +1,96 @@
...
+  let alwaysNode;
+  let isStarted = false;
+  let isDone = false;
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleAnimWrapper.js]: Do we really need 3 variables for keeping up with animation state? It seems like the only thing we want is to prevent `start` and `stop` being called twice.

Also note that since animation runs in native there is a chance that `isDone` may not yet been set to `true` even so the animation has ended in which case using it this way may not be 100% correct

File: src/animations/backwardsCompatibleAnimWrapper.js [hunk 42]
```diff
@@ -0,0 +1,96 @@
...
+                      wasStopped = false;
+                    }
+                  }),
+                  stopClock(newClock),
+                ]),
+                currentState.position,
+              ])
+            )
+          );
+          returnMethod = currentReturnMethod;
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleAnimWrapper.js]: What's the point of having `returnMethod` and `currentReturnMethod`? Looks like if you rename `currentReturnMethod` -> `returnMethod` it would have the same effect

File: src/animations/backwardsCompatibleAnimWrapper.js [hunk 43]
```diff
@@ -0,0 +1,96 @@
...
+          alwaysNode.__addChild(value);
+          value.__setAnimation({
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleAnimWrapper.js]: I'd like for the "private" api calls to be as unfrequent as possible. Would it make sense if add children from within `__setAnimation` method? Especially that we "removeChild" in "detachAnimation" method

File: src/core/AnimatedValue.js [hunk 44]
```diff
@@ -14,21 +16,44 @@ export default class AnimatedValue extends AnimatedNode {
   }
 
   __detach() {
+    this.__detachAnimation();
     super.__detach();
   }
 
+  __detachAnimation(result) {
+    if (this.animation) {
+      this.animation.returnMethod &&
```
Comment (kmagiera) [inline src/core/AnimatedValue.js]: let's call this `callback` or `animationCallback` rather than `returnMethod` (it does not return anything)

File: src/core/AnimatedValue.js [hunk 45]
```diff
@@ -14,21 +16,44 @@ export default class AnimatedValue extends AnimatedNode {
   }
 
   __detach() {
+    this.__detachAnimation();
     super.__detach();
   }
 
+  __detachAnimation(result) {
```
Comment (kmagiera) [inline src/core/AnimatedValue.js]: Looks like we no longer call `detachAnimation` when the animation finishes

File: ios/Nodes/REANode.m (hunk 32 above)
Comment (osdnk) [reply ios/Nodes/REANode.m]: Are you sure, @kmagiera? 
I got "No visible @interface for 'REANode' declares the selector 'update'" as update is part of REAFinal protocol

File: src/animations/backwardsCompatibleAnimWrapper.js (hunk 42 above)
Comment (osdnk) [reply src/animations/backwardsCompatibleAnimWrapper.js]: I wish to have access to method out of the scope, but now it's not needed

File: src/animations/backwardsCompatibleAnimWrapper.js (hunk 41 above)
Comment (osdnk) [reply src/animations/backwardsCompatibleAnimWrapper.js]: `isStarted` is need get info if is method start is triggered for 
```
      if (isStarted) {
        returnMethod && returnMethod({ finished: false });
        return;
      }
```
on start callback

`isDone` is needed not to start animation for the second time without callback and give proper warn

`wasStopped` is needed to give proper callback in `call`  in 48 line

File: src/core/AnimatedValue.js (hunk 45 above)
Comment (osdnk) [reply src/core/AnimatedValue.js]: I can't get it

Comment (osdnk) [reply src/core/AnimatedValue.js]: We call by ` value.__setAnimation(null, !wasStopped);`

File: src/core/AnimatedValue.js [hunk 46]
```diff
@@ -14,9 +16,25 @@ export default class AnimatedValue extends AnimatedNode {
...
+  __detachAnimation(animationCallback) {
+    if (this.animation) {
+      this.animation.returnMethod &&
+        this.animation.returnMethod({ finished: animationCallback });
+      this.animation.node.__removeChild(this);
+    }
+    this.animation = null;
+  }
+
+  __setAnimation(animation, previousAnimationCallback = false) {
```
Comment (kmagiera) [inline src/core/AnimatedValue.js]: rename -> `hasFinishedPrevAnimation`

File: src/animations/backwardsCompatibleAnimWrapper.js [hunk 47]
```diff
@@ -0,0 +1,90 @@
...
+
+function backwardsCompatibleInvoke(node, AnimationClass, value, config) {
+  const newClock = new Clock();
+  const currentState = AnimationClass.getDefaultState();
+  let alwaysNode;
+  let isStarted = false;
+  let isDone = false;
+  let wasStopped = false;
+  return {
+    start: returnMethod => {
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleAnimWrapper.js]: rename `returnMethod` => `animationCallback`

File: src/core/AnimatedValue.js [hunk 48]
```diff
@@ -29,6 +47,11 @@ export default class AnimatedValue extends AnimatedNode {
     this.__forceUpdateCache(value);
   }
 
+  setValue(value) {
+    this.__detachAnimation();
```
Comment (kmagiera) [inline src/core/AnimatedValue.js]: I think when the value is set using imperative API the animation should stop. It seems like calling into "detach" would not actually stop the animation.

File: src/core/AnimatedValue.js [hunk 49]
```diff
@@ -14,9 +16,25 @@ export default class AnimatedValue extends AnimatedNode {
   }
 
   __detach() {
+    this.__detachAnimation(false);
```
Comment (kmagiera) [inline src/core/AnimatedValue.js]: Same comment here as in `setValue()` – it appears this won't actually stop the animation from running, we will just invoke the callback

File: src/animations/backwardsCompatibleAnimWrapper.js [hunk 50]
```diff
@@ -0,0 +1,90 @@
...
+            set(
+              value,
+              block([
+                cond(clockRunning(newClock), 0, startClock(newClock)),
+                node(newClock, currentState, config),
+                cond(currentState.finished, [
+                  call([], () => {
+                    isStarted = false;
+                    isDone = true;
+                    value.__setAnimation(null, !wasStopped);
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleAnimWrapper.js]: don't understand why we can't just call __detachAnimation here?

File: src/animations/backwardsCompatibleAnimWrapper.js [hunk 51]
```diff
@@ -0,0 +1,90 @@
...
+                    }
+                  }),
+                  stopClock(newClock),
+                ]),
+                currentState.position,
+              ])
+            )
+          );
+          value.__setAnimation({
+            node: alwaysNode,
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleAnimWrapper.js]: this seems a bit weird that we crate that object here. It makes it difficult to track who should control calling the callback. Would be easier to follow if we were passing node and callback as two arguments. That would indicate we expect `value` to call the callback.

Alternatively we should just pass the whole "animation" object to the `value` – the object that has `start` and `stop` methods + we can make it export `alwaysNode` for the sake of `value`s access and make it controll calling `animationCallback` when `value` or someone else calls into `stop`

File: src/animations/backwardsCompatibleAnimWrapper.js [hunk 52]
```diff
@@ -0,0 +1,90 @@
...
+        return;
+      }
+      if (!isStarted) {
+        console.warn("Animation hasn't been started");
+        return;
+      }
+      wasStopped = true;
+      evaluateOnce(set(currentState.finished, 1), currentState.finished);
+    },
+    __stopImmediately_testOnly: () => value.__setAnimation(null, true),
```
Comment (kmagiera) [inline src/animations/backwardsCompatibleAnimWrapper.js]: I'd prefer if we use more logic here that is used outside of the test environment. Specifically we could define the method from line 44 in the main scope (next to `alwaysNode`) and here we could call that method simulating that code which is expected to execute when  we set `finished = 1` is run. We should also call to `stop()` here prior to that

File: src/Animated.test.js [hunk 53]
```diff
@@ -0,0 +1,177 @@
...
+const { Value, timing, spring, decay } = Animated;
+describe('Reanimated backward compatible API', () => {
+  beforeEach(() => {
+    let numberOfNodes = 0;
+    ReanimatedModule.createNode = () => numberOfNodes++;
+    ReanimatedModule.dropNode = () => numberOfNodes--;
+    ReanimatedModule.getNumberOfNodes = () => numberOfNodes;
+  });
+
+  const checkIfAttachAndDetachNodesProperly = animation => {
```
Comment (kmagiera) [inline src/Animated.test.js]: naming -> `checkIfNodesGetDetachedCorrectly` ?

File: src/Animated.test.js [hunk 54]
```diff
@@ -0,0 +1,177 @@
...
+      }
+    }
+    const ref = React.createRef();
+    const wrapper = renderer.create(<TestComponent ref={ref} />);
+
+    ref.current.start1();
+    ref.current.start2();
+
+    wrapper.unmount();
+    expect(ReanimatedModule.getNumberOfNodes()).toBe(0);
```
Comment (kmagiera) [inline src/Animated.test.js]: we should check that number of nodes is non zero prior to unmount

File: src/Animated.test.js [hunk 55]
```diff
@@ -0,0 +1,177 @@
...
+        return (
+          <Animated.View style={{ transform: [{ translateX: this.transX }] }} />
+        );
+      }
+    }
+    const ref = React.createRef();
+    const wrapper = renderer.create(<TestComponent ref={ref} />);
+
+    ref.current.start1();
+    ref.current.start2();
```
Comment (kmagiera) [inline src/Animated.test.js]: why two animations instead of one?

File: src/Animated.test.js [hunk 56]
```diff
@@ -0,0 +1,177 @@
...
+    const numberOfNodesBeforeDetach = ReanimatedModule.getNumberOfNodes();
+    wrapper1.unmount();
+    const numberOfNodesAfterDetach = ReanimatedModule.getNumberOfNodes();
+    const result =
+      // 3 means AnimatedProps, AnimatedStyle and AnimatedTransform
+      // which are nodes not related to animation and has to be detached
+      numberOfNodesBeforeDetach - 3 === numberOfNodesAfterDetach &&
+      numberOfNodesAfterDetach > 3;
+    wrapper2.unmount();
+    expect(result && ReanimatedModule.getNumberOfNodes() === 0).toBeTruthy();
```
Comment (kmagiera) [inline src/Animated.test.js]: we should check if result is true immediately, then do another check that nodes are at 0 after second unmount. This way if the test fails it is straightforward to tell which of these two conditions has failed. You can have as many expect statements in a test as you wish

File: src/Animated.test.js (hunk 55 above)
Comment (osdnk) [reply src/Animated.test.js]: I wish to check what happens if two anims affect one node.  Changed name of test

File: src/core/AnimatedValue.js (hunk 48 above)
Comment (osdnk) [reply src/core/AnimatedValue.js]: It actually stops animation 

Please see line 27

File: src/core/AnimatedValue.js (hunk 49 above)
Comment (osdnk) [reply src/core/AnimatedValue.js]: same
